### PR TITLE
Set CUDA12 as default in GPU packages

### PR DESCRIPTION
**Pull request by yf711**
### Description

* Swap cuda version 11.8/12.2 in GPU CIs
* Set CUDA12 as default version in yamls of publishing nuget/python/java GPU packages
* Suppress warnings as errors of flash_api.cc during ort win-build

### CODE REVIEW

File: tools/ci_build/github/azure-pipelines/py-cuda-publishing-pipeline.yml [hunk 1]
```diff
@@ -16,7 +16,7 @@ parameters:
 variables:
   - name: ArtifactFeed
     ${{ if eq(parameters.isReleaseBuild, false) }}:
-      value: ort-cuda-12-nightly
+      value: ORT-Nightly
     ${{ else }}:
       value: onnxruntime-cuda-12
```
Comment (tianleiwu) [inline tools/ci_build/github/azure-pipelines/py-cuda-publishing-pipeline.yml:21]: Maybe add onnxruntime-cuda-11 feed.

Comment (snnn) [reply tools/ci_build/github/azure-pipelines/py-cuda-publishing-pipeline.yml:21]: Then our CPU packages and GPU packages will not be in the same feed.

**Comment by jywu-msft** [comment]
pls fix python lint errors

**Comment by jywu-msft** [comment]
do we know why there are build warning/errors on CUDA build on Windows?  @tianleiwu @jchen351  , was it encountered previously in CUDA12 pipeline?

   215>D:\a\_work\1\s\onnxruntime\contrib_ops\cuda\bert\flash_attention\flash_api.cc(97,27): error C2220: the following warning is treated as an error [D:\a\_work\1\b\RelWithDebInfo\onnxruntime_providers_cuda.vcxproj]
##[warning]onnxruntime\contrib_ops\cuda\bert\flash_attention\flash_api.cc(97,27): Warning C4616: #pragma warning: warning number '2220' not a valid compiler warning
   215>D:\a\_work\1\s\onnxruntime\contrib_ops\cuda\bert\flash_attention\flash_api.cc(97,27): warning C4616: #pragma warning: warning number '2220' not a valid compiler warning [D:\a\_work\1\b\RelWithDebInfo\onnxruntime_providers_cuda.vcxproj]
##[warning]onnxruntime\contrib_ops\cuda\bert\flash_attention\flash_api.cc(99,14): Warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data
   215>D:\a\_work\1\s\onnxruntime\contrib_ops\cuda\bert\flash_attention\flash_api.cc(99,14): warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data [D:\a\_work\1\b\RelWithDebInfo\onnxruntime_providers_cuda.vcxproj]
##[warning]onnxruntime\contrib_ops\cuda\bert\flash_attention\flash_api.cc(100,14): Warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data
   215>

**Comment by snnn** [comment]
Our nightly pipelines do not treat warning as errors. So the problems have been there for a while.

**Comment by tianleiwu** [comment]
> do we know why there are build warning/errors on CUDA build on Windows? @tianleiwu @jchen351 , was it encountered previously in CUDA12 pipeline?

@Alejandro Cid Delgado to fix the warnings in flash_api.cc.

**Comment by snnn** [comment]
Tianlei, the pipeline uses:

python.exe tools\ci_build\build.py --config RelWithDebInfo --build_dir xxxx --skip_submodule_sync --build_csharp --update --parallel --use_binskim_compliant_compile_flags --cmake_generator "Visual Studio 17 2022" --build_shared_lib --enable_onnx_tests --enable_pybind --build_java --build_nodejs --use_cuda --cuda_home=D:\a\_work\_temp\v12.2 --enable_cuda_profiling --enable_transformers_tool_test --cmake_extra_defines CMAKE_CUDA_ARCHITECTURES=86 --cmake_extra_defines onnxruntime_BUILD_UNIT_TESTS=ON --cmake_extra_defines onnxruntime_ENABLE_CUDA_EP_INTERNAL_TESTS=ON

**Comment by snnn** [comment]
The TRT CI pipeline still has CUDA 11 in PATH

D:\a\_work\_temp\TensorRT-10.2.0.19.Windows10.x86_64.cuda-11.8\lib;C:\Users\cloudtest\AppData\Roaming\Python\Python38\Scripts;C:\ToolCache\Python\3.8.10\x64\Scripts;C:\ToolCache\Python\3.8.10\x64;C:\vss-agent\3.241.0\externals\git\cmd;C:\vss-agent\3.241.0\externals\git\mingw64\bin;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\bin\HostX64\x64;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\VC\VCPackages;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\TestWindow;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\TeamFoundation\Team Explorer;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Current\bin\Roslyn;C:\Program Files (x86)\Microsoft SDKs\Windows\v10.0A\bin\NETFX 4.8 Tools\x64\;C:\Program Files (x86)\HTML Help Workshop;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\FSharp\Tools;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Team Tools\DiagnosticsHub\Collector;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\Extensions\Microsoft\CodeCoverage.Console;C:\Program Files (x86)\Windows Kits\10\bin\10.0.22621.0\\x64;C:\Program Files (x86)\Windows Kits\10\bin\\x64;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\\MSBuild\Current\Bin\amd64;C:\Windows\Microsoft.NET\Framework64\v4.0.30319;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\Tools\;C:\Users\cloudtest\AppData\Roaming\Python\Python38\Scripts;C:\ToolCache\Python\3.8.10\x64\Scripts;C:\ToolCache\Python\3.8.10\x64;C:\vss-agent\3.241.0\externals\git\cmd;C:\vss-agent\3.241.0\externals\git\mingw64\bin;C:\ToolCache\Java_Temurin-Hotspot_jdk\11.0.24-8\x64\bin;C:\npm\prefix;C:\ProgramData\Chocolatey\bin;C:\Program Files\Microsoft SDKs\Azure\CLI2\wbin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\dotnet\;C:\Program Files\PowerShell\7\;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn\;C:\Program Files\Microsoft SQL Server\150\Tools\Binn\;C;C:\Program Files\nodejs\;C:\Program Files\CMake\bin;C:\Program Files\Git\cmd;C:\Users\cloudtest\AppData\Local\Microsoft\WindowsApps;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\VC\Linux\bin\ConnectionManagerExe;D:\a\_work\_temp\v11.8\bin;D:\a\_work\_temp\v11.8\extras\CUPTI\lib64;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.8\bin;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.8\extras\CUPTI\lib64

File: tools/ci_build/github/windows/setup_env_cuda.bat [hunk 2]
```diff
@@ -1,17 +1,17 @@
 REM Copyright (c) Microsoft Corporation. All rights reserved.
 REM Licensed under the MIT License.
 
-if exist PATH=%AGENT_TEMPDIRECTORY%\v11.8\ (
```
Comment (snnn) [inline tools/ci_build/github/windows/setup_env_cuda.bat:4]: Now https://github.com/microsoft/onnxruntime/blob/main/tools/ci_build/github/azure-pipelines/templates/jobs/download_win_gpu_library.yml should have already added CUDA to PATH. So this file probably is no longer needed.

Comment (snnn) [reply tools/ci_build/github/windows/setup_env_cuda.bat:4]: #20583

Comment (yf711) [reply tools/ci_build/github/windows/setup_env_cuda.bat:4]: I think this file is still useful to non-pkg CIs like Win GPU/TRT CIs, as this [templates/jobs/download_win_gpu_library.yml](https://github.com/microsoft/onnxruntime/blob/main/tools/ci_build/github/azure-pipelines/templates/jobs/download_win_gpu_library.yml) is only used by templates supporting packaging CIs

**Review by jchen351 — APPROVED** [review]

File: tools/ci_build/github/windows/setup_env_cuda.bat (hunk 2 above)
Comment (snnn) [reply tools/ci_build/github/windows/setup_env_cuda.bat:4]: Yeah, we should consider making them consistent.

Comment (yf711) [reply tools/ci_build/github/windows/setup_env_cuda.bat:4]: I see. Will explore on that after this PR and deprecate batch scripts

**Comment by tianleiwu** [comment]
Could we upgrade cuDNN 8 to cuDNN 9 in all pipelines?

PyTorch 2.4 uses cudnn 9 for both cuda 11.8 and cuda 12.4. I think we can also move to cudnn 9 completely (including for CUDA 11).

**Pull request merged**